Add option to generate:tests for XSD validation on 'merged files'

soumyau opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
https://jira.corp.magento.com/browse/MQE-1541

- Added a --fastdebug (-a) option to generate:tests option to run XSD validation against 'merged files'.
- Introduce a new .env parameter 'MFTF_FAST_DEBUG' that can be used to enable/disable fast debugging
- Logging xsd validation errors as CRITICAL in mftf log

### Fixed Issues (if relevant)
1. Fixed incorrect parsing while reading MFTF_DEBUG parameters from .env


### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/verification tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)
 - [ ] Changes to Framework doesn't have backward incompatible changes for tests or have related Pull Request with fixes to tests

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2-functional-testing-framework?pullRequest=349) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2-functional-testing-framework?pullRequest=349) before we can accept your contribution.<br/>

KevinBKozan commented on src/Magento/FunctionalTestingFramework/Util/Logger/MftfLogger.php:
```diff
@@ -28,4 +28,21 @@ public function deprecation($message, array $context = [])
+     * @param  array  $context The log context.
+     * @return void
+     */
+    public function buildFailure($message, array $context = [])
```
Rename to just `critical` to follow pattern set by `deprecation` function above.

KevinBKozan commented on src/Magento/FunctionalTestingFramework/Util/Logger/MftfLogger.php:
```diff
@@ -28,4 +28,21 @@ public function deprecation($message, array $context = [])
+     */
+    public function buildFailure($message, array $context = [])
+    {
+        $message = "BUILD FAILURE: " . $message;
```
Change to `CRITICAL FAILURE` to keep this build-agnostic

soumyau replied:
fixed

KevinBKozan commented on etc/config/.env.example:
```diff
@@ -42,8 +42,9 @@ BROWSER=chrome
-#*** Bool property which allows the user to toggle debug output during test execution
+#*** Bool properties which allows the user to toggle debug output during test execution
 #MFTF_DEBUG=
+#MFTF_FAST_DEBUG=
```
What else is `MFTF_DEBUG` used for, and why don't we just repurpose the old flag instead of adding a new one?
I don't see the point in adding `FAST_DEBUG` if noone will even want ot use `MFTF_DEBUG` flag to begin with.

soumyau replied:
This flag is removed - enabled merged file debugging for generate:tests by default

KevinBKozan commented on src/Magento/FunctionalTestingFramework/Config/Reader/Filesystem.php:
```diff
@@ -231,8 +231,13 @@ protected function validateSchema($configMerger, $filename = null)
+                        [ "file"=> $filename ? $filename: ":mergedFile:", "error" => $error]
+                    );
+                }
+                throw new \Exception("Error: XSD schema issues found in file(s) " . $filename . "\n");
```
I'm not sure I'm reading this right, but does this exception get thrown as soon as a single merged file fails the schema validation check? Or does `configMerger` iterate through every file?

soumyau replied:
Not changing the current functionality; schema validation check fails at the first error encounter As discussed, will create a new story as this may not be a quick implementation.

soumyau commented:
> Some changes required to logger, but also have a bigger question around the new flag

Changed the implementation as per recent discussions: 1. enabled debugging for generate:tests by default 2.--debug runs developer mode allowing enhanced debugging.3. introduced input option 'ignore' to skip debugging for backward compatibility

KevinBKozan commented on src/Magento/FunctionalTestingFramework/Console/GenerateTestsCommand.php:
```diff
@@ -78,9 +80,8 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $force = $input->getOption('force');
         $time = $input->getOption('time') * 60 * 1000; // convert from minutes to milliseconds
-        $debug = $input->getOption('debug');
+        $mode = $input->getOption('debug');
```
I'm not 100% sure how you can currently go from `--debug` to `self::MODE_DEVELOPER`, I may be missing something. Unless the intention is for me to input `--debug developer` or something, in which case we shoudl change the desc to show the required options.

If not, we can map true/false all existing --debug cases (we'd need to keep it string input though).

`--debug true` == `debug = true` == `self::MODE_DEVELOPER`
`--debug false` == `debug = false ` == `self::MODE_PRODUCTION`
`no debug flag` == `debug = null` == `self:MODE_DEFAULT`

soumyau replied:
Updated as per recent discussions
- in GenerateTestsCommand `$debug` input option is set to optional; 
`--debug` == LEVEL_DEVELOPER (backward compatibility)
`--debug developer` == LEVEL_DEVELOPER
`--debug none` == DEBUG_NONE
`no debug flag` == LEVEL_DEFAULT

KevinBKozan commented on dev/tests/verification/TestModule/Test/BasicFunctionalTest.xml:
```diff
@@ -24,7 +24,7 @@
         <comment stepKey="basicCommentWithNoData" userInput="{{emptyData.noData}}"/>
         <comment stepKey="basicCommentWithDefinitelyNoData" userInput="{{emptyData.definitelyNoData}}"/>
-        <comment stepKey="ReadinessCheckSkipped" userInput="skipReadiness" skipReadiness="true"/>
+        <comment stepKey="basicCommentWithData" userInput="seeComment"/>
```
Change is unrelated to PR, not sure why it's here.

KevinBKozan commented on dev/tests/verification/TestModule/ActionGroup/BasicActionGroup.xml:
```diff
@@ -115,10 +115,6 @@
         <see selector=".selector" userInput="{{sameStepKeyAsArg}}" stepKey="arg1" />
     </actionGroup>
 
-    <actionGroup name="actionGroupWithSkipReadinessActions">
```
Change is unrelated to PR, not sure why it's here.

KevinBKozan commented on dev/tests/verification/Resources/BasicFunctionalTest.txt:
```diff
@@ -60,9 +60,7 @@ class BasicFunctionalTestCest
-		$I->skipReadinessCheck(true);
-		$I->comment("skipReadiness");
-		$I->skipReadinessCheck(false);
+		$I->comment("seeComment");
```
Change is unrelated to PR, not sure why it's here.

KevinBKozan commented on dev/tests/verification/Tests/ActionGroupGenerationTest.php:
```diff
@@ -185,17 +185,6 @@ public function testActionGroupWithArgContainingStepKey()
         $this->generateAndCompareTest('ActionGroupContainsStepKeyInArgText');
     }
 
-    /**
```
Change is unrelated to PR, not sure why it's here.

KevinBKozan commented:
Spoke in person with @soumyau. Decision was as follows:
* Keep `--debug` as optional
* Map `--debug` to `developer` for backwards compatibility
* Allow setting of debug level via `--debug developer/production/default`

soumyau commented on src/Magento/FunctionalTestingFramework/Config/MftfApplicationConfig.php:
```diff
@@ -14,6 +14,14 @@ class MftfApplicationConfig
+     */
+    const MODE_DEFAULT = "default";
+    const MODE_DEVELOPER = "developer";
+    const MODE_PRODUCTION = "off";
```
- changed to debug levels: `default`, `developer`, `none` as per recent discussions.
- in GenerateTestsCommand `$debug` input option is set to optional; 
`--debug` == LEVEL_DEVELOPER (backward compatibility)
`--debug developer` == LEVEL_DEVELOPER
`--debug none` == DEBUG_NONE
`no debug flag` == LEVEL_DEFAULT

soumyau commented on src/Magento/FunctionalTestingFramework/Config/MftfApplicationConfig.php:
```diff
@@ -14,6 +14,14 @@ class MftfApplicationConfig
+     */
+    const LEVEL_DEFAULT = "default";
+    const LEVEL_DEVELOPER = "developer";
+    const DEBUG_NONE = "none";
```
done

soumyau commented on src/Magento/FunctionalTestingFramework/Config/MftfApplicationConfig.php:
```diff
@@ -72,7 +80,15 @@ private function __construct(
+                $this->debugLevel = $debugLevel;
+                break;
+            default:
+                $this->debugLevel = self::LEVEL_DEVELOPER;
```
defaulting to `developer` for backward compatibility, if the user doesn't pass any option with `--debug` it will run the original debug mode i.e developer; also if any random value is passed. Alternatively, could add a `null` case for just `--debug` and have `default` to `LEVEL_DEFAULT` and throw exception to the user for an incorrect command.

KevinBKozan commented on dev/tests/functional/tests/MFTF/DevDocs/Test/DevDocsTest.xml:
```diff
@@ -19,6 +19,6 @@
 
         <amOnPage stepKey="openMFTFDevDocPage" url="{{MFTFDocPage.url}}" />
-        <see stepKey="verifyPageIntroText" selector="{{contentSection.pageIntro}}" userInput="Introduction to the Magento Functional Testing Framework" />
+        <see stepKey="verifyPageIntroText" selector="{{contentSection.pageIntro}}" userInput="Introduction to the Magento Functional Testing Framework" skipReadiness=""/>
```
Line was introduced by accident, pretty sure.

soumyau replied:
Fixed
